storage-service/README.md

```diff
@@ -23,39 +24,116 @@ To ensure all necessary environment variables are set correctly this section des
+The way the document upload event is saved on the blockchain decides where the document can be found.
 
 ![architecture_with_external_document_storage](./doc/images/architecture_with_external_document_storage.JPG)
 
 ### Document distribution
 
-When documents are shared between multiple nodes (organizations), the particular nodes gets the decrypted keys of the documents. Then, the documents from the other node can be accessed by the storage-service through the API, which is exposed by the frontend (nginx).
+The enabled storage-service is used in Trubudget to prevent a node from sharing its documents without any restriction among the blockchain network. This way documents are stored in a s3 storage connected to the storage-service and not on-chain. On the blockchain only document ids, access-secrets and document download URLs are stored but not the actual document data.
+To share the document among the organizations in the network a special mechanism is used to only give organizations access which really need the document access.
+
+#### Distribution mechanism
+
+The access of the documents are restricted by a secret generated by the storage-service of the storage the document is saved in. To download a document from a storage-service three parts of information have to be known:
+
+- Valid secret
+- Document ID
+- URL of storage-service
+  These information are found on-chain. The document ID and the URL of the storage-service are shared between all nodes and can be read by any node in the network. The valid secret is saved encrypted instead and can only be accessed by specific organizations.
+  To encrypt and decrypt this secret, RSA keypairs (public and private keys) are used. Every node in the network needs to share it's public key before connecting to a network. This is done on api startup.
+  If a document should be uploaded for an organization, which is currently defined by the permissions `workflowitem.view`, the shared public key of this organization is used to encrypt the document's secret. The secret can only be read by the organization which provides the right private key.
+
+But where is the private/public key pair coming from?
+An organization creates it's own private key automatically to sign each transaction as organization. This private key is shared encrypted with the `ORGANIZATION_VAULT_SECRET` in the organization-stream(`org:KfW`) on-chain. This private key is used to generate the public key. This public key is shared via the `public_keys` stream.
+
+How does Trubudget perform a document upload including the storage-service ?
+
+1. The api has to be started with the external available URL of the storage-service - `STORAGE_SERVICE_EXTERNAL_URL`
+1. A document upload is triggered via one of the intents `workflowitem.create` or `workflowitem.update`
+1. The document is stored via storage-service onto the connected storage
+1. A secret and ID are generated and returned by the storage-service if the document was successfully stored
```

Comment:
the secret is generated by the storage service and the id of the document is generated by the api during `workflowitem.create` or `workflowitem.update`